src/common/scene-helpers.c

```diff
@@ -107,6 +107,14 @@ lab_wlr_scene_output_commit(struct wlr_scene_output *scene_output,
 		return false;
 	}
 
+	if (state->tearing_page_flip) {
+		if (!wlr_output_test_state(wlr_output, state)) {
+			wlr_log(WLR_DEBUG, "Output test for tearing failed on %s, "
+				"trying page-flip without tearing", wlr_output->name);
+			state->tearing_page_flip = false;
```

Comment:
In case the test commit returns `true` but the actual commit fails we should likely have another safeguard in place to retry the commit with tearing disabled. I don't know if that is a possible scenario for any GPU driver but if it is we would skip the frame completely with this PR.

Reply:
Is that also the case when `state == &output->pending`?

Reply:
Yes, that issue would exist with both states. We could refactor this slightly to only have a single `wlr_output_commit_state()` and merge the state reset in `output_state_commit()` from `src/output-state.c` into `lab_wlr_scene_output_commit()` and remove `output_state_commit()` to have less duplication. E.g. something like (completely untested pseudo code):
```c
lab_wlr_scene_output_commit([..]) {
	[..]

	if (state->buffer && is_magnify_on()) {
		magnify(output, state->buffer, &additional_damage);
	}

	if (!wlr_output_commit_state(wlr_output, state)) {
		if (state->tearing_page_flip) {
			/*
			 * Test commit with tearing went through but actual
			 * commit failed, lets try again without tearing
			 */
			state->tearing_page_flip = false;
			if (wlr_output_commit_state(wlr_output, state)) {
				goto commit_ok;
			}
		}
		wlr_log(WLR_INFO, "Failed to commit output %s", wlr_output->name);
		return false;
	}

commit_ok:
	if (state == &output->pending) {
		wlr_output_state_finish(&output->pending);
		wlr_output_state_init(&output->pending);
	}

	if (!wlr_box_empty(&additional_damage)) {
		[..]
	}
	return true;
}
```

CC @jlindgren90

Reply:
To see it was a bit confusing why there were 2 different functions for commit used, with almost identical names (for someone pretty new to LabWC and wlroots at least). 

So guessing from `state == &output->pending` and `wlr_output_state_finish` being called on the state it is the commit for the final output state? The function name might need some more context

Reply:
I think something like this would be the cleanest;

```c
bool committed = false;
if (state == &output->pending) {
  committed = output_state_commit(output);
} else {
  committed = wlr_output_commit_state(wlr_output, state);
}
/* 
 * Handle case where the ouput state test for tearing succeeded, 
 * but actual commit failed. Retry wihout tearing.
 */
if (!committed && state->tearing_page_flip) {
  state->tearing_page_flip = false;
  committed = wlr_output_commit_state(wlr_output, state);
}
if (!committed) {
  wlr_log(WLR_INFO, "Failed to commit output %s", wlr_output->name);
  return false;
}
```

Or we could use the conditional operator if you prefer that;

```c
bool committed = (state == &output->pending)
       ? output_state_commit(output)
       : wlr_output_commit_state(wlr_output, state);
/* 
 * Handle case where the ouput state test for tearing succeeded, 
 * but actual commit failed. Retry wihout tearing.
 */
if (!committed && state->tearing_page_flip) {
  state->tearing_page_flip = false;
  committed = wlr_output_commit_state(wlr_output, state);
}
if (!committed) {
  wlr_log(WLR_INFO, "Failed to commit output %s", wlr_output->name);
  return false;
}
```

Reply:
This would not reset the shared `output->pending` state on non-tearing retry.
We should simply get rid of `output_state_commit()` in this function and reset `output->pending` directly in case that is what we are committing.

 > To me it was a bit confusing why there are 2 different functions for commit used, with almost identical names (for someone pretty new to LabWC and wlroots at least).

Historical reason, needs to be cleaned up.

> So guessing from `state == &output->pending` and `wlr_output_state_finish` being called on the state it is the commit for the final output state? The function name might need some more context

No, we do have a shared pending state for the next commit in `output->pending`. If a commit went through we need to reset it (finish -> init) so we don't end up committing the same things over and over again. Some parts of the code base do use their own state instead of the shared `output->pending` one in which case we must not reset the shared state because we would loose some uncommitted changes.

Reply:
Changed, should cover all cases now